server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,493 @@
+          mkInvocation se 1000 decodedHeaders errMsg []
+        HParse _ detail -> do
+          let errMsg = TBS.fromLBS $ J.encode detail
+          mkInvocation se 1001 decodedHeaders errMsg []
+        HStatus errResp -> do
+          let respPayload = hrsBody errResp
+              respHeaders = hrsHeaders errResp
+              respStatus = hrsStatus errResp
+          mkInvocation se respStatus decodedHeaders respPayload respHeaders
+        HOther detail -> do
+          let errMsg = (TBS.fromLBS $ J.encode detail)
+          mkInvocation se 500 decodedHeaders errMsg []
+  liftExceptTIO $
+    Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) $ do
+    insertInvocation invocation
+    retryOrMarkError se err
+
+retryOrMarkError :: ScheduledEventFull -> HTTPErr a -> Q.TxE QErr ()
+retryOrMarkError se@ScheduledEventFull {..} err = do
+  let mRetryHeader = getRetryAfterHeaderFromHTTPErr err
+      mRetryHeaderSeconds = parseRetryHeaderValue =<< mRetryHeader
+      triesExhausted = sefTries >= strcNumRetries sefRetryConf
+      noRetryHeader = isNothing mRetryHeaderSeconds
+  if triesExhausted && noRetryHeader
+    then do
+      markError
+    else do
+      currentTime <- liftIO getCurrentTime
+      let delay = fromMaybe (round $ unNonNegativeDiffTime
+                             $ strcRetryIntervalSeconds sefRetryConf)
+                    $ mRetryHeaderSeconds
+          diff = fromIntegral delay
+          retryTime = addUTCTime diff currentTime
+      setRetry se retryTime
+  where
+    markError =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+        UPDATE hdb_catalog.hdb_scheduled_events
+        SET error = 't', locked = 'f'
+        WHERE id = $1
+      |] (Identity sefId) True
+
+{- Note [Scheduled event lifecycle]
+
+A scheduled event can be in one of the five following states at any time:
+
+1. Delivered
+2. Cancelled
+3. Error
+4. Locked
+5. Dead
+
+A scheduled event is marked as delivered when the scheduled event is processed
+successfully.
+
+A scheduled event is marked as error when while processing the scheduled event
+the webhook returns an error and the retries have exhausted (user configurable)
+it's marked as error.
+
+A scheduled event will be in the locked state when the graphql-engine fetches it
+from the database to process it. After processing the event, the graphql-engine
+will unlock it. This state is used to prevent multiple graphql-engine instances
+running on the same database to process the same event concurrently.
+
+A scheduled event will be marked as dead, when the difference between the
+current time and the scheduled time is greater than the tolerance of the event.
+
+A scheduled event will be in the cancelled state, if the `cancel_scheduled_event`
+API is called against a particular scheduled event.
+
+The graphql-engine will not consider those events which have been delivered,
+cancelled, marked as error or in the dead state to process.
+-}
```

Comment:
This comment is incredibly helpful, thank you. 🙂 I do have a few minor followup comments:

1. You write that an event can be in one of those five states, but I think you are missing one state: the state of events that have been created, but haven’t been processed at all. Maybe you could call that state “scheduled.”

2. It’s hard to immediately grok from this description which states are actually accessible from which other states. One easy way to solve that is with a flowchart:

   ```
   {- Note [Scheduled event lifecycle]
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   Scheduled events move between six different states over the course of their
   lifetime, as represented by the following flowchart:

     ┌───────────┐    ┌────────┐    ┌───────────┐
     │ scheduled │───→│ locked │───→│ delivered │
     └───────────┘    └────────┘    └───────────┘
                            │       ┌───────────┐
                            ├──────→│ cancelled │
                            │       └───────────┘
                            │       ┌───────┐
                            ├──────→│ error │
                            │       └───────┘
                            │       ┌──────┐
                            └──────→│ dead │
                                    └──────┘
   ```

3. As the comment explains, these states are all mutually-exclusive: an event can only be in one of them at a time. But in the SQL database, they’re stored as multiple separate boolean flags. This seems bad to me—we would never model something that way in Haskell! We wouldn’t write this:

   ```haskell
   -- | Note: only one of these should ever be 'True' at a time.
   data EventStatus = EventStatus
     { esLocked    :: Bool
     , esDelivered :: Bool
     , esCancelled :: Bool
     , esError     :: Bool
     , esDead      :: Bool
     }
   ```

   We’d write this:

   ```haskell
   data EventStatus
     = ESScheduled
     | ESLocked
     | ESDelivered
     | ESCancelled
     | ESError
     | ESDead
   ```

   SQL doesn’t have sum types, but it *can* model simple enumerations like this just fine. Why not have a single `status` column, like this:

   ```sql
   CREATE TABLE hdb_catalog.hdb_scheduled_events (
     id TEXT DEFAULT gen_random_uuid() UNIQUE,
     name TEXT,
     scheduled_time TIMESTAMPTZ NOT NULL,
     status TEXT NOT NULL DEFAULT 'scheduled'
       CONSTRAINT valid_status (status IN ('scheduled', 'locked', 'delivered',
                                           'cancelled', 'error', 'dead')),
     additional_payload JSON,
     tries INTEGER NOT NULL DEFAULT 0,
     next_retry_at TIMESTAMPTZ,
     created_at TIMESTAMP DEFAULT NOW(),

     PRIMARY KEY (name, scheduled_time),
     FOREIGN KEY (name) REFERENCES hdb_catalog.hdb_scheduled_trigger(name)
       ON UPDATE CASCADE ON DELETE CASCADE
   );

   CREATE INDEX hdb_scheduled_event_status ON hdb_scheduled_events (status);
   ```

   This should be just as performant given the index on the `status` column. (In fact, the lack of an index on these columns in the current data model is probably bad!)

4. It seems like maybe the flowchart is actually missing one transition, since I think an event can go from `locked` back to `scheduled` if it retries. I also realized while writing this comment that my original flowchart was wrong, since an event actually goes directly from `scheduled` to `cancelled`.

Now having read and understood all that context, here’s my suggested tweaked version of your (quite excellent!) comment:

```suggestion
{- Note [Scheduled event lifecycle]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Scheduled events move between six different states over the course of their
lifetime, as represented by the following flowchart:

  ┌───────────┐      ┌────────┐      ┌───────────┐
  │ scheduled │─(a)─→│ locked │─(b)─→│ delivered │
  └───────────┘      └────────┘      └───────────┘
      │    ↑              │          ┌───────┐
      │    └────(c)───────┼─────(d)─→│ error │
      │                   │          └───────┘
      │                   │          ┌──────┐
      │                   └─────(e)─→│ dead │
      │                              └──────┘
      │                              ┌───────────┐
      └─────────────────────────(f)─→│ cancelled │
                                     └───────────┘

When an event is first created, it starts in the 'scheduled' state, and it can
transition to other states in the following ways:

  a. When graphql-engine fetches an event from the database to process it, it
     sets its state to 'locked'. This prevents multiple graphql-engine instances
     running on the same database from processing the same event concurrently.

  b. When a scheduled event is processed successfully, it is marked 'delivered'.

  c. If an event fails to be processed, but it hasn’t yet reached its maximum
     retry limit, its retry counter is incremented and it is returned to the
     'scheduled' state.

  d. If an event fails to be processed and *has* reached its retry limit, its
     state is set to 'error'.

  e. If for whatever reason the difference between the current time and the
     scheduled time is greater than the tolerance of the event, it will not be
     processed and its state will be set to 'dead'.

  f. Finally, the `cancel_scheduled_event` API moves an unprocessed event
     directly to the 'canceled' state.
-}
```